src/dialog/llm/lcel_default.py

```diff
@@ -0,0 +1,113 @@
+    def memory(self) -> BaseChatMemory:
+        if self.session_id:
+            return generate_memory_instance(
+                session_id=self.session_id,
+                parent_session_id=self.parent_session_id
+            )
+        return None
+
+    def generate_prompt(self, text):
+        self.relevant_contents = get_most_relevant_contents_from_message(
+            text, top=LLM_RELEVANT_CONTENTS, dataset=self.dataset
+        )
+        prompt_config = self.config.get("prompt")
+        fallback = prompt_config.get("fallback") or self.config.get("fallback").get("prompt") # maintaining compatibility with the previous configuration
+        header = prompt_config.get("header")
+        suggested = prompt_config.get("suggested")
+        messages = []
+        if len(self.relevant_contents) > 0:
+            context = "Context: \n".join(
+                [f"{c.question}\n{c.content}\n" for c in self.relevant_contents]
+            )
+            self.prompt = ChatPromptTemplate.from_template(
+                "\n".join([
+                    SystemMessagePromptTemplate.from_template(header),
+                    messages.append(SystemMessagePromptTemplate.from_template(
+                        f"{suggested}. {context}"
+                    )),
+                    "{chat_history}",
+                    "{user_message}"
+                ])
+            )
+        else:
+            ChatPromptTemplate.from_template(
+                "\n".join([
+                    fallback,
+                    HumanMessagePromptTemplate.from_template("{user_message}")
+                ])
+            )
+        self.prompt = ChatPromptTemplate.from_messages(messages)
+
+        if VERBOSE_LLM:
+            logging.info(f"Verbose LLM prompt: {self.prompt.pretty_print()}")
+
+    @property
+    def llm(self) -> LLMChain:
+        llm_config = self.config.get("model", {})
+        conversation_options = {
+            "llm": ChatOpenAI(
+                **llm_config,
+                openai_api_key=self.llm_key or OPENAI_API_KEY
+            ),
+            "prompt": self.prompt,
+            "verbose": self.config.get("verbose", False)
+        }
+
+        if self.memory:
+            buffer_config = {
+                "chat_memory": self.memory,
+                "memory_key": "chat_history",
+                "return_messages": True,
+                "k": self.config.get("memory_size", LLM_MEMORY_SIZE)
+            }
+            conversation_options["memory"] = ConversationBufferWindowMemory(
+                **buffer_config
+            )
+
+        return LLMChain(**conversation_options)
+
+    def postprocess(self, output: str) -> str:
+        if self.memory:
+            asyncio.create_task(categorize_conversation_history(self.memory))
+        return output
+
+    def process(self, input):
+        processed_input = self.preprocess(input)
+        self.generate_prompt(processed_input)
+        chain = self.prompt | self.llm | self.parser
```

Comment:
I left a comment about this PR in [the issue](https://github.com/talkdai/dialog/issues/124#issuecomment-1986872974), but I'll paste here as well:

This is a good first step towards adopting LCEL, since it has small changes, but it **does not close the issue**, I think we should discuss better in the issue (where I gave several ideas and it i'd be good to hear other ideas) how the roadmap could be to really get a LCEL interface which langchain's users will easily understand.

For example, I think that AbstractLLM forces a flow which langchain's chains already were designed for (the idea of input data, retrieval phase, prompt generation, branches and if-else cases, llm calls, post processing and output parsers), and, I my opinion, to really make dialog flexible with LCEL, the roadmap should go in the direction of removing this custom class in favor of building chains entirely with langchain's objects in every step (runnables), since this is what guarantees new functionalities like streaming, async, parallelism, retries/fallbacks, access to intermediate results, and tracing with langsmith etc. In this direction, I imagine dialog exposing in the api any chain the user wants to build or even combine, as long as input/output remain the same meaning, with a default chain available (using runnables in all steps). To really make a new class which follows LCEL will demand more changes and probably several PRs.

Reply:
@lgabs we are building chains into classes because they are:

1) easily inheritable and modifiable - every developer knows a little bit of OOP and our focus here is to make LLMs very accessible to users;

2) we are doing small changes so we dont break retro-compatibility with existing code;

3) because the user is supposed to implement their own LLMs using whichever methods they want, so giving them a class with methods that allow them to pre/post-process and run the LLM are quicker.

If a user wants to use a different approach to the LLMs, they are allowed to modify and do whatever they want with their own version of the LLM or even modify the source code for talkd.